src/modules/verifiable-credential/verifiable-credential.controller.ts

```diff
@@ -0,0 +1,83 @@
+import {
+  Body,
+  Controller,
+  Get,
+  Post,
+  Query,
+  UseInterceptors,
+} from '@nestjs/common';
+import { ApiTags, ApiOkResponse } from '@nestjs/swagger';
+import { User } from '../../common/user.decorator';
+import { Auth } from '../auth/auth.decorator';
+import { SentryErrorInterceptor } from '../interceptors/sentry-error-interceptor';
+import { Logger } from '../logger/logger.service';
+import { GetCredentialsDto } from './dtos/get-credential.dto.';
+import { PresentationDefinitionDto } from './dtos/presentation-definition.dto';
+import { VerifiableCredentialDto } from './dtos/verifiable-credential.dto';
+import { VCStorage } from './entities/vc-storage.entity';
+import { VCMatchService } from './services/vc-match.service';
+import { VCStorageService } from './services/vc-storage.service';
+
+@Auth()
+@UseInterceptors(SentryErrorInterceptor)
+@Controller({ path: 'vc', version: '1' })
+@ApiTags('Verifiable Credential')
+export class VerifiableCredentialController {
+  constructor(
+    private readonly logger: Logger,
+    private readonly vcStorageService: VCStorageService,
+    private readonly vcMatchService: VCMatchService
+  ) {
+    this.logger.setContext(VerifiableCredentialController.name);
+  }
+
+  @Get('/match')
+  public async getMatchedCredentialsForVpDefinition(
+    @User() user: string,
+    @Query() presentationDefinition: PresentationDefinitionDto
+  ) {
+    return await this.vcMatchService.matchCredentials(
+      presentationDefinition,
+      user
+    );
+  }
+
+  @ApiOkResponse({
+    type: [VerifiableCredentialDto],
+    description:
+      'Returns all verifiable credentials related to the user. If subject is provided, returns only verifiable credentials related to the subject. If id is provided, returns only verifiable credential with the provided id. Default is all verifiable credentials.',
+  })
+  @Get('/storage')
+  public async getCredentialsById(
+    @User() user: string,
+    @Query() { subject, id }: GetCredentialsDto
+  ) {
+    let vcItems: VCStorage[];
+
+    if (subject) {
+      vcItems = await this.vcStorageService.findBySubject(subject, user);
+    } else if (id) {
+      const vcItem = await this.vcStorageService.findById(id, user);
+      vcItems = [vcItem];
+    } else {
+      vcItems = await this.vcStorageService.findAll(user);
+    }
+
+    return vcItems.map((vcItem) => vcItem.vc);
+  }
+
+  @Post('/storage')
+  public async saveCredentials(
+    @Body() vc: VerifiableCredentialDto,
+    @User() user: string
+  ) {
+    await this.vcStorageService.save({
+      vcId: vc.id,
+      issuer: vc.issuer,
+      subject: vc.credentialSubject.id,
+      holder: user,
+      vc,
+    });
+    this.logger.log(`Saved verifiable credential ${vc.id}`);
+  }
+}
```

Comment:
Instead of this custom REST interface, I think we might as well try to use the DIF Decentralized Web Node interface (or another standard storage interface if you know one), even if it will take a little bit more time to implement. I will make a brief PR today to sketch out what this could look like